src/lib/utils/lib.ts

```diff
@@ -52,3 +55,109 @@ export const getTransactionReceipt = async (
 
 export const isDefined = <T>(val: T | null | undefined): val is T =>
   typeof val !== 'undefined' && val !== null
+
+export const getBlockRangesForL1Block = async ({
+  targetL1BlockNumber,
+  provider,
+}: {
+  targetL1BlockNumber: number
+  provider: JsonRpcProvider
+}) => {
+  const arbitrumProvider = new ArbitrumProvider(provider)
+  const currentArbBlock = await arbitrumProvider.getBlockNumber()
+
+  const arbitrumChainId = (await arbitrumProvider.getNetwork()).chainId
+  const { nitroGenesisBlock } = l2Networks[arbitrumChainId]
+
+  // Define the starting point to be closer to the current block for efficiency.
+  let startArbBlock = Math.floor(currentArbBlock * 0.95)
+  let endArbBlock = currentArbBlock
+
+  async function getL1Block(forL2Block: number) {
+    const l2Block = Math.max(nitroGenesisBlock, forL2Block)
+    const { l1BlockNumber } = await arbitrumProvider.getBlock(l2Block)
+    return l1BlockNumber
+  }
+
+  // Binary search to find the starting Arbitrum block that corresponds to the L1 block number.
+  async function getL2StartBlock() {
+    let result
+    let start = startArbBlock
+    let end = endArbBlock
+
+    while (start <= end) {
+      // Calculate the midpoint of the current range.
+      const mid = start + Math.floor((end - start) / 2)
```

Comment:
I am a little bit unsure about these casting between BigNumber and number, it should be fine and I used it somewhere else in the sdk too.